**Modelo de Preaprobacion -Naibe Bayes**
Base de datos de Solicitantes de Credito mediante el modelo Naive Bayes.
0. se procede con la carga de las librerias de trabajo.

In [1]:
import numpy as np #Funciones Matematicas.
import pandas as pd #Comunicacion con excel.
from sklearn.naive_bayes import GaussianNB #Modelo de Clasificacion - Funcion Gaussiana.
from sklearn.metrics import confusion_matrix #Matriz de Confusion - La metrica de Medida.

1. Secargan los dato.


In [17]:
nxl = '/content/SolicitantesCreditoNaiveBayes.xlsx'
XDB=pd.read_excel(nxl, sheet_name=2) #Traer los datos de la hoja AAA
XDB.dropna() #Eliminar datos faltantes, celdas vacias

#Seleccionamos las variables de interes.
XDB=XDB[['Edad','Ingresos','Egresos','Cuota (COP)','PreApr']]

#Datos de Entrada
XD=np.array(XDB[['Edad','Ingresos','Egresos','Cuota (COP)']])
#Datos de Salida - Valor deseado - Lo que quiero que aprenda el modelo
yd=np.array(XDB['PreApr'])


2. Implementacion del Modelo

In [49]:
mnb=GaussianNB() #Llamamos el modelo
mnb.fit(XD,yd) #Entrenamos el modelo
ydp=mnb.predict(XD) #Muestra los datos de salida que el obtuvo

#Parametros del modelo por categoria - PreApr/PreNeg
u=mnb.theta_ #Se estiman las medidas
sigma=np.sqrt(mnb.var_) #Se estiman las desviaciones

#Las vaariaciones de las variables
LI=u-sigma
LS=u+sigma

print("La media de las variables por categoria: ")
print(u)
print("El limite inferior de la variación de las variables por categoria es:")
print(LI)
print("El limite superior de la variación de las variables por categoria es:")
print(LS)

#Metrica de evaluación - Como se comporto el modelo identificando la PreApr-PreNeg
cm=confusion_matrix(yd,ydp)
print("La matriz de confusion es:")
print(cm)

La media de las variables por categoria: 
[[  40.98411626  431.42680644  401.54995644   16.88529983]
 [  41.10752688 1029.98947251  900.19312144   43.3269839 ]]
El limite inferior de la variación de las variables por categoria es:
[[ 28.77108358 105.62889147  58.77709557 -12.88412769]
 [ 28.85646196 557.55604444 374.75974307 -21.65436704]]
El limite superior de la variación de las variables por categoria es:
[[  53.19714893  757.22472141  744.32281731   46.65472735]
 [  53.35859181 1502.42290058 1425.62649981  108.30833484]]
La matriz de confusion es:
[[2563  396]
 [1192 1691]]


3. Se procede con la evaluacion del modelo.


In [47]:
VN=cm[0,0] #Verdaderos Negativos
FP=cm[0,1] #Falsos Positivos
FN=cm[1,0] #Falsos Negativos
VP=cm[1,1] #Verdaderos Positivos

Exactitud=(VP+VN)/len(XD)
print("La exactitud del modelo es:", Exactitud) #Comportamiento General

TE=(FN+FP)/len(XD)
print("La tasa de error del modelo es:", TE) #Porcentaje de Equivocaciones

Sensibilidad=VP/(VP+FP)
print("La sensibilidad del modelo es:", Sensibilidad) #Como se comporto frente a los positivos

Especificidad=VN/(VN+FN)
print("La especificidad del modelo es:", Especificidad) #Como se comporto frente a los negativos

Precision=VP/(VP+FN)
print("La precición del modelo es:", Precision) #Como le pego a los clasificados positivos

PredNeg=VN/(VN+FP)
print("La probabilidad de Negativo del modelo es:", PredNeg) #Como le pego a los clasificados negativos



La exactitud del modelo es: 0.7281752824375214
La tasa de error del modelo es: 0.2718247175624786
La sensibilidad del modelo es: 0.8102539530426449
La especificidad del modelo es: 0.6825565912117177
La precición del modelo es: 0.5865417967395075
La probabilidad de Negativo del modelo es: 0.8661710037174721


4. Que pasa con una sola persona?

In [50]:
Xun=np.array([[38,1356,1686,18]]) #Persona especifica del ejemplo de excel
ydun=mnb.predict(Xun) #
print(ydun)

[1]


Analisis de Resultados:
De los resultados arrojados por el modelo se puede observar que los preaprobados en general presentan edades (), ingresos (), egresos () y cuotas () mayores a los prenegados.

Con respecto a las metricas se puede observar que el modelo funicono mejor clasificacndo los positivos, como lo dice el indicador de sensibilidad el cual alcanzo el 81% de certeza.

De manera general se puede observar que el modelo logro una exactitud que se ubico por encima del 70% en promedio, y una tasa de error cercana al 30% en promedio.

De manera general esta exactitud lo convierte en un buen modelo. Con respecto al solicitante del credito del ejemplo del excel este pertenece a la categoria de aprobado.